File: 二叉树/145.cpp
#include <vector>
#include <stack>
#include <iostream>
using namespace std;

//Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// 递归
class Recursion
{
public:
    vector<int> res;
    vector<int> postorderTraversal(TreeNode *root)
    {
        if (root)
        {
            postorderTraversal(root->left);
            postorderTraversal(root->right);
            res.push_back(root->val);
        }
        return res;
    }
};

// 迭代
class Iteration
{
public:
    vector<int> postorderTraversal(TreeNode *root)
    {
        vector<int> res;
        stack<TreeNode *> NodeStack;
        NodeStack.push(root);
        while (!NodeStack.empty())
        {
            TreeNode *node = NodeStack.top();
            res.push_back(node->val);
            NodeStack.pop();
            if (node->left)
                NodeStack.push(node->left);
            if (node->right)
                NodeStack.push(node->right);
        }
        reverse(res.begin(),res.end());
        return res;
    }
};

int main()
{
    TreeNode *root = new TreeNode(0);
    root->left = new TreeNode(1);
    root->left->left = new TreeNode(3);
    root->left->right = new TreeNode(4);
    root->right = new TreeNode(2);
    root->right->left = new TreeNode(5);
    root->right->right = new TreeNode(6);
    Iteration obj;
    vector<int> res = obj.postorderTraversal(root);
    for (auto num : res)
        cout << num << endl;
}

File: 二叉树/145.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# 递归
class Recursion:
    def postorderTraversal(self, root: TreeNode) -> list:
        res = []
        self.postorder(root, res)
        return res

    def postorder(self, root, res):
        if(root):
            self.postorder(root.left, res)
            self.postorder(root.right, res)
            res.append(root.val)

# 迭代
class Iteration:
    def postorderTraversal(self, root: TreeNode) -> list:
        res = []
        stack = []
        if(root):
            stack.append(root)
        while(stack):
            node = stack.pop()
            res.append(node.val)
            if(node.left):
                stack.append(node.left)
            if(node.right):
                stack.append(node.right)
        res =res[::-1]

        return res


root = TreeNode(0)
root.left = TreeNode(1)
root.right = TreeNode(2)
root.left.left = TreeNode(3)
root.left.right = TreeNode(4)
root.right.left = TreeNode(5)
root.right.right = TreeNode(6)
obj = Iteration()
print(obj.postorderTraversal(root))